res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,2696 @@
+            // is used in TT mode
+            stackAvg: [],
+            stackAvgIdx: 0,
+            input: function(value) {
+                const oldValue = this.oldValue;
+                this.oldValue = value;
+                if (oldValue === null) {
+                    // This is to avoid the issue where the first time, we diff with 0, leading to the absolute value
+                    return;
+                }
+                let diff = value - oldValue;
+
+                if (diff > wheelRelativeMax / 2) {
+                    diff = (wheelRelativeMax - value + oldValue) * -1;
+                } else if (diff < -1 * (wheelRelativeMax / 2)) {
+                    diff = wheelRelativeMax - oldValue + value;
+                }
+
+                this.stack[this.stackIdx] = diff / wheelTimerDelta;
+                this.stackIdx = (this.stackIdx + 1) % 5;
+
+                const avgSpeed = (this.stack.reduce((ps, v) => ps + v, 0) / this.stack.length) * wheelTicksPerTimerTicksToRevolutionsPerSecond;
+
+                this.stackAvg[this.stackAvgIdx] = avgSpeed;
+                this.stackAvgIdx = (this.stackAvgIdx + 1) % 40;
+
+                const ttAvgSpeed = this.stackAvg.reduce((ps, v) => ps + v, 0) / this.stackAvg.length;
+
+                if (avgSpeed === 0 &&
+                    engine.getValue(this.group, "scratch2") === 0 &&
+                    engine.getValue(this.group, "jog") === 0 &&
+                    this.deck.wheelMode !== wheelModes.motor) {
+                    return;
+                }
+
+                switch (this.deck.wheelMode) {
+                case wheelModes.motor:
+                    // console.log(this.group+"\t"+ttAvgSpeed / baseRevolutionsPerSecond);
+                    engine.setValue(this.group, "scratch2", ttAvgSpeed / baseRevolutionsPerSecond);
+                    break;
+                case wheelModes.loopIn:
+                    engine.setValue(
+                        this.group,
+                        "loop_start_position",
+                        Math.min(
+                            engine.getValue(
+                                this.group,
+                                "loop_start_position"
+                            ) + (avgSpeed * loopWheelMoveFactor),
+                            engine.getValue(
+                                this.group,
+                                "loop_end_position"
+                            ) - loopWheelMoveFactor
+                        )
+                    );
+                    break;
+                case wheelModes.loopOut:
+                    engine.setValue(
+                        this.group,
+                        "loop_end_position",
+                        engine.getValue(
+                            this.group,
+                            "loop_end_position"
+                        ) + (avgSpeed * loopWheelMoveFactor)
+                    );
```

Comment:
This code is difficult to read. I suggest assigning variables to the `engine.getValue` results so the math can fit on one line comfortably.